# HG changeset patch
# User Massimiliano Fasi <massimiliano.fasi@gmail.com>
# Date 1404814777 -7200
#      Tue Jul 08 12:19:37 2014 +0200
# Node ID 9addb5ad942650d5995cf10c5c7b20c635fa3206
# Parent  93805f9256a94500dce3c2c0d8eadd270e575f4e
bandwidth.m: Add new function.

* NEWS: Announce new function.

* scripts/linear-algebra/module.mk: New function added to the build system.

* scripts/help/__unimplemented__.m: New function removed from unimplemented.

* scripts/linear-algebra/isbanded.m: New function added as documentation reference.

* scripts/linear-algebra/isdiag.m: New function added as documentation reference.

* scripts/linear-algebra/istril.m: New function added as documentation reference.

* scripts/linear-algebra/istriu.m: New function added as documentation reference.

* scripts/linear-algebra/bandwidth.m: New function.

* linalg.txi: Add bandwidth function to manual.

diff --git a/NEWS b/NEWS
--- a/NEWS
+++ b/NEWS
@@ -51,23 +51,20 @@ Summary of important user-visible change
  ** linkprop has been completely re-coded for performance and Matlab
     compatibility.  It now returns a linkprop object which must be stored
     in a variable for as long as the graphic objects should remain linked.
     To unlink properties use 'clear hlink' where hlink is the variable
     containing the linkprop object.
 
  ** Other new functions added in 4.2:
 
-      dir_in_loadpath     javachk                      
-      hgload              linkaxes
-      hgsave              numfields
-      istril  
-      istriu   
-      isdiag  
-      isbanded
+      bandwidth            isbanded        javachk
+      dir_in_loadpath      isdiag          linkaxes
+      hgload               istril          numfields
+      hgsave               istriu
 
  ** Deprecated functions.
 
     The following functions have been deprecated in Octave 4.2 and will
     be removed from Octave 4.6 (or whatever version is the second major
     release after 4.2):
 
       bicubic
diff --git a/doc/interpreter/linalg.txi b/doc/interpreter/linalg.txi
--- a/doc/interpreter/linalg.txi
+++ b/doc/interpreter/linalg.txi
@@ -91,16 +91,18 @@ Hermitian.  However, the attempt to fact
 detect a non-Hermitian matrix.
 
 @node Basic Matrix Functions
 @section Basic Matrix Functions
 @cindex matrix functions, basic
 
 @DOCSTRING(balance)
 
+@DOCSTRING(bandwidth)
+
 @DOCSTRING(cond)
 
 @DOCSTRING(det)
 
 @DOCSTRING(eig)
 
 @DOCSTRING(givens)
 
diff --git a/scripts/help/__unimplemented__.m b/scripts/help/__unimplemented__.m
--- a/scripts/help/__unimplemented__.m
+++ b/scripts/help/__unimplemented__.m
@@ -98,17 +98,17 @@ function txt = __unimplemented__ (fcn)
     case {"xlsread", "xlsfinfo", "xlswrite", "wk1read", "wk1finfo", "wk1write"}
       txt = ["Functions for spreadsheet style I/O ", ...
              "(.xls .xlsx .sxc .ods .dbf .wk1 etc.) " , ...
              "are provided in the io package. ", ...
              "See @url{http://octave.sf.net/io/}."];
 
     ## control system
     case {"absorbDelay", "allmargin", "append", "augstate", "balreal", ...
-          "balred", "balredOptions", "bandwidth", "bdschur", "bode", ...
+          "balred", "balredOptions", "bdschur", "bode", ...
           "bodemag", "bodeoptions", "bodeplot", "c2d", "c2dOptions", ...
           "canon", "care", "chgFreqUnit", "chgTimeUnit", "connect", ...
           "connectOptions", "covar", "ctrb", "ctrbf", "ctrlpref", "d2c", ...
           "d2cOptions", "d2d", "d2dOptions", "damp", "dare", "db2mag", ...
           "dcgain", "delay2z", "delayss", "dlqr", "dlyap", "dlyapchol", ...
           "drss", "dsort", "dss", "dssdata", "esort", "estim", "evalfr", ...
           "feedback", "filt", "frd", "frdata", "freqresp", "gcare", "gdare", ...
           "genfrd", "genmat", "gensig", "genss", "get", "getBlockValue", ...
diff --git a/scripts/linear-algebra/bandwidth.m b/scripts/linear-algebra/bandwidth.m
new file mode 100644
--- /dev/null
+++ b/scripts/linear-algebra/bandwidth.m
@@ -0,0 +1,110 @@
+## Copyright (C) 2014 Massimiliano Fasi
+##
+## This file is part of Octave.
+##
+## Octave is free software; you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 3 of the License, or (at
+## your option) any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+## General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the file COPYING.  If not, see
+## <http://www.gnu.org/licenses/>.
+
+## -*- texinfo -*-
+## @deftypefn  {Function File} {@var{bw} =} bandwidth (@var{A}, @var{type})
+## @deftypefnx {Function File} {[@var{lower}, @var{upper}] =} bandwidth (@var{A})
+## Compute the bandwidth of @var{A}.
+##
+## The @var{type} argument is the string @qcode{"lower"} for the lower
+## bandwidth and @qcode{"upper"} for the upper bandwidth.
+## If no @var{type} is specified return both the lower and upper bandwidth
+## of @var{A}. 
+##
+## The lower/upper bandwidth of a matrix is the number of
+## subdiagonals/superdiagonals with nonzero entries.
+## 
+## @seealso{isbanded, isdiag, istril, istriu}
+## @end deftypefn
+
+## Author: Massimiliano Fasi
+
+function [lower, upper] = bandwidth (A, type)
+
+  if (! ((nargin == 1 && nargout == 2) || (nargin == 2 && nargout <= 1)))
+    print_usage ();
+  endif
+
+  if (! isnumeric (A) && ! islogical (A) || ndims (A) != 2)
+    error ("bandwidth: A must be a 2-D numeric or logical matrix");
+  elseif (nargin == 2 && ! (strcmp (type, "lower") || strcmp (type, "upper")))
+    error ('bandwidth: TYPE must be "lower" or "upper"');
+  endif
+
+  if (nargin == 1)
+    [i, j] = find (A);
+    if (isempty (i))
+      lower = upper = 0;
+    else
+      lower = max (i - j);
+      upper = max (j - i);
+    endif
+  else
+    [i, j] = find (A);
+    if (isempty (i))
+      lower = 0;
+    elseif (strcmp (type, "lower"))
+      lower = max (i - j);
+    else
+      lower = max (j - i);
+    endif
+  endif
+
+endfunction
+
+
+%!test
+%! [a,b] = bandwidth (speye (100));
+%! assert ([a,b] == [0,0]);
+%! assert (bandwidth (speye (100), "upper"), 0);
+%! assert (bandwidth (speye (100), "lower"), 0);
+
+%!test
+%! A = [2 3 0 0 0; 1 2 3 0 0; 0 1 2 3 0; 0 0 1 2 3; 0 0 0 1 2];
+%! [a,b] = bandwidth (A);
+%! assert ([a,b] == [1,1]);
+%! assert (bandwidth (A, "lower"), 1);
+%! assert (bandwidth (A, "upper"), 1);
+
+%!assert (bandwidth ([], "lower"), 0)
+%!assert (bandwidth ([], "upper"), 0)
+%!assert (bandwidth (zeros (3,3), "lower"), 0)
+%!assert (bandwidth (zeros (3,3), "upper"), 0)
+%!assert (bandwidth (ones (5,5), "lower"), 4)
+%!assert (bandwidth (ones (5,5), "upper"), 4)
+
+%!test
+%! [a,b] = bandwidth ([]);
+%! assert ([a,b] == [0,0]);
+%!test
+%! [a,b] = bandwidth (zeros (3,3));
+%! assert ([a,b] == [0,0]);
+%!test
+%! [a,b] = bandwidth (ones (5,5));
+%! assert ([a,b] == [4,4]);
+
+%% Test input validation
+%!error bandwidth ()
+%!error bandwidth (1,2,3)
+%!error [a,b,c] = bandwidth (ones (2))
+%!error [a,b] = bandwidth (ones (2), "upper")
+%!error <A must be a 2-D numeric or logical> bandwidth ("string", "lower")
+%!error <A must be a 2-D numeric or logical> bandwidth (ones (3,3,3), "lower")
+%!error <TYPE must be "lower" or "upper"> bandwidth (ones (2), "uper")
+%!error <TYPE must be "lower" or "upper"> bandwidth (ones (2), "uppper")
+
diff --git a/scripts/linear-algebra/isbanded.m b/scripts/linear-algebra/isbanded.m
--- a/scripts/linear-algebra/isbanded.m
+++ b/scripts/linear-algebra/isbanded.m
@@ -18,17 +18,17 @@
 
 ## -*- texinfo -*-
 ## @deftypefn {Function File} {} isbanded (@var{x}, @var{lower}, @var{upper})
 ## Return true if @var{x} is a matrix with entries confined between
 ## @var{lower} diagonals below the main diagonal and @var{upper} diagonals
 ## above the main diagonal.
 ##
 ## @var{lower} and @var{upper} must be non-negative integers.
-## @seealso{isdiag, istril, istriu}
+## @seealso{isdiag, istril, istriu, bandwidth}
 ## @end deftypefn
 
 ## Author: Massimiliano Fasi
 
 function retval = isbanded (A, lower, upper)
 
   if (nargin != 3)
     print_usage ();
diff --git a/scripts/linear-algebra/isdiag.m b/scripts/linear-algebra/isdiag.m
--- a/scripts/linear-algebra/isdiag.m
+++ b/scripts/linear-algebra/isdiag.m
@@ -14,17 +14,17 @@
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with Octave; see the file COPYING.  If not, see
 ## <http://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
 ## @deftypefn {Function File} {} isdiag (@var{x})
 ## Return true if @var{x} is a diagonal matrix.
-## @seealso{isbanded, istril, istriu, diag}
+## @seealso{isbanded, istril, istriu, diag, bandwidth}
 ## @end deftypefn
 
 ## Author: Massimiliano Fasi
 
 function retval = isdiag (A)
 
   if (nargin != 1)
     print_usage ();
diff --git a/scripts/linear-algebra/istril.m b/scripts/linear-algebra/istril.m
--- a/scripts/linear-algebra/istril.m
+++ b/scripts/linear-algebra/istril.m
@@ -17,17 +17,17 @@
 ## <http://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
 ## @deftypefn {Function File} {} istril (@var{x})
 ## Return true if @var{x} is a lower triangular matrix.
 ##
 ## A lower triangular matrix has nonzero entries only on the main diagonal
 ## and below.
-## @seealso{istriu, isbanded, isdiag, tril}
+## @seealso{istriu, isbanded, isdiag, tril, bandwidth}
 ## @end deftypefn
 
 ## Author: Massimiliano Fasi
 
 function retval = istril (A)
 
   if (nargin != 1)
     print_usage ();
diff --git a/scripts/linear-algebra/istriu.m b/scripts/linear-algebra/istriu.m
--- a/scripts/linear-algebra/istriu.m
+++ b/scripts/linear-algebra/istriu.m
@@ -17,17 +17,17 @@
 ## <http://www.gnu.org/licenses/>.
 
 ## -*- texinfo -*-
 ## @deftypefn {Function File} {} istriu (@var{x})
 ## Return true if @var{x} is an upper triangular matrix.
 ##
 ## An upper triangular matrix has nonzero entries only on the main diagonal
 ## and above.
-## @seealso{isdiag, isbanded, istril, triu}
+## @seealso{isdiag, isbanded, istril, triu, bandwidth}
 ## @end deftypefn
 
 ## Author: Massimiliano Fasi
 
 function retval = istriu (A)
 
   if (nargin != 1)
     print_usage ();
diff --git a/scripts/linear-algebra/module.mk b/scripts/linear-algebra/module.mk
--- a/scripts/linear-algebra/module.mk
+++ b/scripts/linear-algebra/module.mk
@@ -1,11 +1,12 @@
 FCN_FILE_DIRS += linear-algebra
 
 linear_algebra_FCN_FILES = \
+  linear-algebra/bandwidth.m \
   linear-algebra/commutation_matrix.m \
   linear-algebra/cond.m \
   linear-algebra/condest.m \
   linear-algebra/cross.m \
   linear-algebra/duplication_matrix.m \
   linear-algebra/expm.m \
   linear-algebra/housh.m \
   linear-algebra/isbanded.m \
